### DISCUSSION TITLE
Using git-sync to keep a user code deployment in sync with a git branch

### DISCUSSION THREAD

**Discussion by craustin** [dagster-io/dagster#18060] [Q&A]
Is it possible / a good idea to use git-sync to keep a k8s-based “user code deployment” in-sync with a ‘prod’ branch in the repo? It’s handy in Airflow to commit and have DAG changes automatically synced up to the running airflow web server/scheduler in a few seconds. I’d rather avoid building and deploying a new docker image when I update Dagster code location code. Would it work to just update the code in the container a-la git-sync and reload the code location? How do others do this?

**Comment by alangenfeld** [comment]
Theoretically a set-up like this can work ya, here are some details for making deployed code servers reloadable
https://github.com/dagster-io/dagster/discussions/14709

**Reply by craustin** [reply]
Thanks @alangenfeld. I've gotten this (almost) working by forking the helm charts, adding support for `git-sync`, and using `codeServerArgs` to make the code location reloadable.

Something I hadn't considered though: When using `K8sRunLauncher`, the pods that execute the run will _also_ need to have a `git-sync` init container to pull down the Dagster defs since that pod will need to reconstruct the job definition. When using the Helm charts, the `K8sRunLauncher` config is in the _dagster_ instance config and not in the _user code deployment_ chart config. This is awkward since it means the `git-sync` volume/mount/container specs need to be included in both the dagster and dagster user code deployment charts. If there was a way to configure the `K8sRunLauncher` in the code deployment yaml, that'd simplify this.

**Comment by mst** [comment]
+1 
I think this would be useful too. I don't understand why we'd want to go the detour about rebuilding a image and redeploy the container for changes.

**Comment by dduong1603** [comment]
how would you do that though (beside patching the deployment)? I'm trying to do something similar, but I don't see the code location chart having config to accept additional containers/sidecars...(I'm a newbie when it comes to k8s though so I probably miss something 🤣)

**Comment by andrescolodrero** [comment]
This is obviously quite a good feature and will allow more gitops friendly deployments.

After deploy Dagster with Helm, it creates a gRPC server, and Dagster Daemon get the configuration code to load from a specific server.

load_from:
  - grpc_server:
      host: k8s-example-user-code-1
      port: 3030
      location_name: k8s-example-user-code-1

The question if the grpc server can be configured to download the code from a git repo. Then the daemon doesnt load from git, but from the gRPC server.

**Comment by the-xentropy** [comment]
This was actually fairly easy. Only took a few hours - and I have it dynamically installing my python requirements too.

(Using the dagster kubernetes helm chart)

0. Configure the deployment to use a podSecurityContext setting the fsGroup to e.g. 1000. Your cluster might need runAsUser and runAsGroup too, but just set them all to the same value.
1. Add git-sync as a sidecarContainer, pulling in code to a known location. If you don't do the above you'll get file permission issues since git-sync will write with unusable permissions.
2. Add a startupProbe to the git-sync sidecarContainer which checks for the existence of the definitions file (I do `stat <package root - I.E. below the first absolute import>/somepackagename/__init__.py`). Basically we want a long-lived init container that pulls in the code and delays the normal container start-up for until it has successfully synced the code at least once.
3. Set the dagster user-code-example gRPC command line options to:
```
          dagsterApiGrpcArgs:
            - "--working-directory"
            - "<package root - I.E. below the first absolute import>"
            - "--python-file"
            - "<package root - I.E. below the first absolute import>/somepackagename/__init__.py"
```
In my case the package root is the github repo's root folder, where .git is. My `__init__.py` imports its own code using e.g. `from somepackagename import <blah>` - relative imports won't work well.

For the python requirements, I just added code that runs `python pip --target <package root - I.E. below the first absolute import> -r <package root - I.E. below the first absolute import>/requirements.txt` before the rest of my imports since the working-directory flag means that the python interpreter will find the packages as long as I download dependencies there.

For those with more static dependencies, it's probably easier to just build a derivative image of the user-code-example and pip install your requirements then.

That said, having done this - the only real benefit I'm getting from this above doing an automatic build+push on every github commit is that I don't have to manually restart the deployment when I have new code (and worry about image caching..) and I don't have to bake in dependencies into the image and can hot-swap them as needed (I install them lazily when/if they're needed, so most pip installs don't run).


So going back to the original premise of this discussion:

1. Yes, it's possible.
2. TBD if it's a good idea. It's working fine for now, but it's not impossible I run into issues with long-living imports, or subtle errors due to cached objects.
